psc-publish/Main.hs

```diff
@@ -0,0 +1,230 @@
+  case mapMaybe D.parseVersion' vs of
+    []  -> userError TagMustBeCheckedOut
+    [x] -> return x
+    xs  -> userError (AmbiguousVersions xs)
+  where
+  clean =
+    trimWhitespace >>> dropPrefix "v"
+  trimWhitespace =
+    dropWhile isSpace >>> reverse >>> dropWhile isSpace >>> reverse
+  dropPrefix prefix str =
+    fromMaybe str (stripPrefix prefix str)
+
+getBowerInfo :: PackageName -> PrepareM (D.GithubUser, D.GithubRepo)
+getBowerInfo pkgName = do
+  r' <- liftIO (catch (Right <$> get (bowerPackageUrl pkgName))
+                      (return . Left))
+  r <- either handleError return r'
+
+  case getUrl r of
+    Right url ->
+      maybe (userError NotOnGithub) return (extractGithub url)
+    Left err ->
+      internalError (JSONError FromBowerApi err)
+
+  where
+  handleError (StatusCodeException status _ _)
+    | status == notFound404 = userError (NotOnBower pkgName)
+  handleError e = otherError (HttpExceptionThrown e)
+
+  getUrl r =
+    parse (key "url" asString)
+          (fromMaybe "" (r ^? responseBody))
+
+bowerDomainName :: String
+bowerDomainName = "bower.herokuapp.com"
+
+bowerPackageUrl :: PackageName -> String
+bowerPackageUrl name =
+  concat [ "https://"
+         , bowerDomainName
+         , "/packages/"
+         , runPackageName name
+         ]
+
+extractGithub :: String -> Maybe (D.GithubUser, D.GithubRepo)
+extractGithub =
+  stripPrefix "git://github.com/"
+   >>> fmap (splitOn "/")
+   >=> takeTwo
+   >>> fmap (D.GithubUser *** (D.GithubRepo . dropDotGit))
+
+  where
+  takeTwo :: [a] -> Maybe (a, a)
+  takeTwo [x, y] = Just (x, y)
+  takeTwo _ = Nothing
+
+  dropDotGit :: String -> String
+  dropDotGit str
+    | ".git" `isSuffixOf` str = take (length str - 4) str
+    | otherwise = str
+
+readProcess' :: String -> [String] -> String -> PrepareM String
+readProcess' prog args stdin = do
+  out <- liftIO (catch (Right <$> readProcess prog args stdin)
+                       (return . Left))
+  either (otherError . ProcessFailed prog args) return out
+
+-- Go through all bower dependencies which contain purescript code, and
+-- extract their versions.
+--
+-- In the case where a bower dependency is taken from a particular version,
+-- that's easy; take that version. In any other case (eg, a branch, or a commit
+-- sha) we print a warning that documentation links will not work, and avoid
+-- linking to documentation for any types from that package.
+--
+-- The rationale for this is: people will prefer to use a released version
+-- where possible. If they are not using a released version, then this is
+-- probably for a reason. However, docs are only ever available for released
+-- versions. Therefore there will probably be no version of the docs which is
+-- appropriate to link to, and we should omit links.
+getResolvedDependencies :: PrepareM [(PackageName, Version)]
+getResolvedDependencies = do
+  pathsBS <- fromString <$> readProcess' "bower" ["list", "--paths", "--json"] ""
```

Comment:
On second thoughts, we can get all the information we need from `bower list --json`.

Using `bower list --json` also allows us to easily identify bower dependencies which haven't been installed, because Bower includes `"missing": true` in the output where appropriate. So we can fail with a nice error if the user forgot to `bower install`.